### DISCUSSION TITLE
Setting up a reset password website

### DISCUSSION THREAD

**Discussion by InfiniteVerma** [docker-mailserver/docker-mailserver#2814] [Q&A] [answered]
There's a request to set up a **reset password website** through which users can directly reset their password instead of asking the mail server administrator. Feature-wise, I can build and deploy a web server that interacts directly with the docker data of ```docker-mailserver``` 

There are the steps I've thought of:

1. User enters their current email and password
2. Backend validates this by directly reading docker data.
3. User enters the new password and confirms
4. Newly entered password gets updated using ```setup.sh``` (To be safe, I can also restart the mail server docker)

I would like to know if the above is valid. And if so, I would like if someone could point out on what files I need to read/update users' passwords.

Thank You!

**Comment by polarathene** [comment, marked as answer]
You may be interested in https://github.com/docker-mailserver/docker-mailserver-admin

It looks like development stalled there however.

---

I worked on cleaning up our `setup` scripts for account management earlier in the year ([_`setup email update` script_](https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/bin/updatemailuser)). This would probably be the better way to update passwords infrequently, which is what you plan to do with step 4.

There shouldn't be a need to restart, the `check-for-changes.sh` script should be monitoring for changes to the relevant files and handle any restarting needed for you. It should be ok, so long as you don't have passwords being changed often as each time those restarts happens it could make the mail-server services unavailable for around 10-20 seconds (depends how long it takes to restart a service).

---

> I would like if someone could point out on what files I need to read/update users' passwords.

The `setup email update` and other account related management for `setup.sh` is handled [here](https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/scripts/helpers/database/manage/postfix-accounts.sh), which works with the file "database" manager scripts I wrote [here](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/scripts/helpers/database/db.sh).

You should be able to perform your updates via those scripts, and that is likely going to be more reliable across releases, rather than directly manipulating any of the "database" files that we manage and generally assume no one else writes to (which could risk corruption).

You will want to perform validation prior to calling the update, as you mention in step 2. The account data is available at `postfix-accounts.cf` file and you an read the password hash from that like we have show in this test:

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/test/setup-cli.bats#L72-L99

Specifically of interest to you might be:

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/test/setup-cli.bats#L95-L96

That `doveadm` command can be provided the current password hash and the plaintext password to do the comparison (_it reads the dovecot config we have for what password scheme to use, which should be `SHA512-CRYPT` at present_). Our accounts data should be in sync with Dovecot (_it has it's own db copy, separate from the `postfix-accounts.cf` we use_), and I think I've got better tests to reference on this topic, I just haven't published the branch for a PR related to it :sweat_smile:
